linalg/sundials.cpp

```diff
@@ -1518,13 +1523,14 @@ ARKStepSolver::ARKStepSolver(MPI_Comm comm, Type type)
 }
 #endif
 
-void ARKStepSolver::Init(TimeDependentOperator &f_)
+void ARKStepSolver::Init(ARKStepODE &f_ark_)
 {
-   // Initialize the base class
-   ODESolver::Init(f_);
+   this->f_arkstep = &f_ark_;
+
+   this->f = dynamic_cast<TimeDependentOperator*>(&f_ark_);
```

Comment:
I think this is the most painful part of such design :face_with_diagonal_mouth: . It is not allowed to dynamic cast to an independent class. The compiler might do that, but it is illegal by the standard.